bot/exts/moderation/silence.py

```diff
@@ -93,87 +93,222 @@ async def _async_init(self) -> None:
-            overwrite = channel.overwrites_for(self._verified_role)
-            if overwrite.send_messages is False or overwrite.add_reactions is False:
-                await channel.send(MSG_UNSILENCE_MANUAL)
+            if isinstance(channel, VoiceChannel):
+                overwrite = channel.overwrites_for(self._verified_voice_role)
+                manual = overwrite.speak is False
+            else:
+                overwrite = channel.overwrites_for(self._verified_msg_role)
+                manual = overwrite.send_messages is False or overwrite.add_reactions is False
+
+            # Send fail message to muted channel or voice chat channel, and invocation channel
+            if manual:
+                await self.send_message(MSG_UNSILENCE_MANUAL, msg_channel, channel)
             else:
-                await channel.send(MSG_UNSILENCE_FAIL)
+                await self.send_message(MSG_UNSILENCE_FAIL, msg_channel, channel)
+
         else:
-            await channel.send(MSG_UNSILENCE_SUCCESS)
+            # Send success message to muted channel or voice chat channel, and invocation channel
+            if isinstance(channel, VoiceChannel):
+                await self._force_voice_sync(channel)
+
+            await self.send_message(MSG_UNSILENCE_SUCCESS, msg_channel, channel, True)
 
-    async def _set_silence_overwrites(self, channel: TextChannel) -> bool:
+    async def _set_silence_overwrites(self, channel: Union[TextChannel, VoiceChannel], kick: bool = False) -> bool:
         """Set silence permission overwrites for `channel` and return True if successful."""
-        overwrite = channel.overwrites_for(self._verified_role)
-        prev_overwrites = dict(send_messages=overwrite.send_messages, add_reactions=overwrite.add_reactions)
+        if isinstance(channel, TextChannel):
+            overwrite = channel.overwrites_for(self._verified_msg_role)
+            prev_overwrites = dict(send_messages=overwrite.send_messages, add_reactions=overwrite.add_reactions)
+        else:
+            overwrite = channel.overwrites_for(self._verified_voice_role)
+            prev_overwrites = dict(speak=overwrite.speak)
+            if kick:
+                prev_overwrites.update(connect=overwrite.connect)
 
         if channel.id in self.scheduler or all(val is False for val in prev_overwrites.values()):
             return False
 
-        overwrite.update(send_messages=False, add_reactions=False)
-        await channel.set_permissions(self._verified_role, overwrite=overwrite)
+        if isinstance(channel, TextChannel):
+            overwrite.update(send_messages=False, add_reactions=False)
+            await channel.set_permissions(self._verified_msg_role, overwrite=overwrite)
+        else:
+            overwrite.update(speak=False)
+            if kick:
+                overwrite.update(connect=False)
+
+            await channel.set_permissions(self._verified_voice_role, overwrite=overwrite)
```

Comment:
How would you get the overwrite object in the if and else branches without calling the `overwrites_for` first? Since it's different for both branches, you'll need to have duplicate code inside the branches or have two checks (current implementation).

The code of the bottom branches (line 238-247) can be refactored into the top branches (226-233), but then the call on 235/236 has to be duplicated.

Reply:
I missed that, sorry. The `overwrites_for` will just have to be duplicated in the if and else blocks at the top.

Reply:
That's a much neater implementation, thank you.